lib/libvfio-user.c

```diff
@@ -890,101 +871,143 @@ process_request(vfu_ctx_t *vfu_ctx)
             if (ret == 0) {
                 _iovecs[1].iov_base = &irq_info;
                 _iovecs[1].iov_len = sizeof irq_info;
-                iovecs = _iovecs;
-                nr_iovecs = 2;
+                *iovecs = _iovecs;
+                *nr_iovecs = 2;
             }
             break;
         case VFIO_USER_DEVICE_SET_IRQS:
-            ret = handle_device_set_irqs(vfu_ctx, hdr.msg_size, fds, nr_fds,
+            ret = handle_device_set_irqs(vfu_ctx, hdr->msg_size, fds, *nr_fds,
                                          cmd_data);
             break;
         case VFIO_USER_REGION_READ:
         case VFIO_USER_REGION_WRITE:
-            iovecs = _iovecs;
-            ret = handle_region_access(vfu_ctx, hdr.msg_size, hdr.cmd,
-                                       &iovecs[1].iov_base, &iovecs[1].iov_len,
+            *iovecs = _iovecs;
+            ret = handle_region_access(vfu_ctx, hdr->msg_size, hdr->cmd,
+                                       &(*iovecs)[1].iov_base,
+                                       &(*iovecs)[1].iov_len,
                                        cmd_data);
-            nr_iovecs = 2;
+            *nr_iovecs = 2;
             break;
         case VFIO_USER_DEVICE_RESET:
             ret = handle_device_reset(vfu_ctx);
             break;
         case VFIO_USER_DIRTY_PAGES:
             // FIXME: don't allow migration calls if migration == NULL
-            ret = handle_dirty_pages(vfu_ctx, hdr.msg_size, &iovecs, &nr_iovecs,
+            ret = handle_dirty_pages(vfu_ctx, hdr->msg_size, iovecs, nr_iovecs,
                                      cmd_data);
             if (ret >= 0) {
-                free_iovec_data = false;
+                *free_iovec_data = false;
             }
             break;
         default:
-            vfu_log(vfu_ctx, VFU_ERR, "bad command %d", hdr.cmd);
+            vfu_log(vfu_ctx, VFU_ERR, "bad command %d", hdr->cmd);
             ret = -EINVAL;
             goto reply;
     }
-
 reply:
+    free(cmd_data);
+    return ret;
+}
+
+static int
+process_request(vfu_ctx_t *vfu_ctx)
+{
+    struct vfio_user_header hdr = { 0, };
+    int ret;
+    int *fds = NULL;
+    int nr_fds;
+    struct iovec _iovecs[2] = { { 0, } };
+    struct iovec *iovecs = NULL;
+    size_t nr_iovecs = 0;
+    bool free_iovec_data = true;
+
+    assert(vfu_ctx != NULL);
+
+    if (device_is_stopped(vfu_ctx->migration)) {
+        return -ESHUTDOWN;
+    }
+
+    /*
+     * FIXME if migration device state is VFIO_DEVICE_STATE_STOP then only
+     * migration-related operations should execute. However, some operations
+     * are harmless (e.g. get region info). At the minimum we should fail
+     * accesses to device regions other than the migration region. I'd expect
+     * DMA unmap and get dirty pages to be required even in the stop-and-copy
+     * state.
+     */
+
+    nr_fds = vfu_ctx->client_max_fds;
+    fds = alloca(nr_fds * sizeof(int));
```

Comment:
I know this is currently safe, but alloca() is the equivalent of strcpy() - you see it and you immediately wonder if this code is secure. I think we should ban it.

In this case, simply doing:

    int fds[MUSER_MAX_CLIENT_FDS] = { [ 0 ... MUSER_MAX_CLIENT_FDS - 1] = -1 };

seems fine, and is *much more clearly* secure. (We want the -1 init to make it clear there's no fd there)

I know you didn't add this, but we should clean it up, if not now.